src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -151,6 +155,59 @@ static BoundNode unwrapListElement(BoundCollectionExpression node, BoundNode ele
             }
         }
 
+        // If we have something like `List<int> l = [.. someEnumerable]`
+        // try rewrite it using `Enumerable.ToList` member if possible
+        private bool TryRewriteSingleElementSpreadToList(BoundCollectionExpression node, TypeWithAnnotations listElementType, [NotNullWhen(true)] out BoundExpression? result)
+        {
+            result = null;
+
+            if (node.Elements is not [BoundCollectionExpressionSpreadElement singleSpread])
+            {
+                return false;
+            }
+
+            if (!ShouldUseAddRangeOrToListMethod(singleSpread.Expression.Type!, singleSpread.EnumeratorInfoOpt?.GetEnumeratorInfo.Method, listElementType))
+            {
+                return false;
+            }
+
+            if (TryGetWellKnownTypeMember(node.Syntax, WellKnownMember.System_Linq_Enumerable__ToList, out MethodSymbol? toListGeneric, isOptional: true))
+            {
+                var toListOfElementType = toListGeneric.Construct([listElementType]);
+                result = _factory.Call(receiver: null, toListOfElementType, singleSpread.Expression);
+                return true;
+            }
+
+            return false;
+        }
+
+        private bool ShouldUseAddRangeOrToListMethod(TypeSymbol spreadType, MethodSymbol? getEnumeratorMethod, TypeWithAnnotations elementType)
+        {
+            var iEnumerableOfTType = _compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerable_T);
+            var iEnumerableOfElementType = iEnumerableOfTType.Construct([elementType]);
+
+            var discardedUseSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+
+            var conversion = _compilation.Conversions.ClassifyImplicitConversionFromType(spreadType, iEnumerableOfElementType, ref discardedUseSiteInfo);
```

Comment:
>var conversion = _compilation.Conversions.ClassifyImplicitConversionFromType(spreadType, iEnumerableOfElementType, ref discardedUseSiteInfo); [](http://example.com/codeflow?start=12&length=141)

Let's use the code from `tryOptimizeSpreadElement` which checks for `ICollection<T>`, then checks the conversion to `addRangeMethod.Parameters[0].Type` rather than the conversion to `IEnumerable<T>` directly. But perhaps we should use `ClassifyImplicitConversionFromType()` rather than `ClassifyConversionFromType()` for the parameter conversion check. Also, the comment in the `ICollection<T>` check from `tryOptimizeSpreadElement` would be good to keep. Obviously, the references to `AddRange` should be updated though.

Reply:
Also, the `ICollection<T>` check in `tryOptimizeSpreadElement` used `spreadElement.EnumeratorInfo.CollectionType` and `spreadElement.EnumeratorInfo.ElementType`. With this change, we are using `spreadElement.Expression.Type` and the element type from the target type. Is there a difference between those pairs of types?

Reply:
> Is there a difference between those pairs of types?

I didn't implement logic in the binding layer, that fills in `EnumeratorInfoOpt`, but from the sanity point of view they _should_ be the same. I preffer avoiding `EnumeratorInfoOpt` since it can be `null`. Or maybe we can assume, that at the point of lowering we always have it and drop additional `null` checks?